### ISSUE TITLE
A quest for better docs for MatExpr and Mat

### ISSUE THREAD

**Issue by howtobeahacker** [Itseez/opencv#5410]
Hi devs,
After using Opencv for my project, I feel that we need a quest to improve docs of MatExpr. There is potential bugs if we do not use the library correctly. 
For example, MatExpr 
```
        // assuming matrix B of type CV_8U, I just want to normalize it into [0, 1]
        Mat A = B/255.0f;  // potiential bug
```
The results are completely different from what we want. The matrix ```A``` is still in CV_8U, so it will lose floating points value to convert to just *0* or *1*. 
My opinion is that we should say that the matrix type of resultant matrix is the same as with the most compatible operand matrix.

**Comment by Aravind-Suresh** [comment]
I am interested to work on this issue to start with. Can someone tell me what is to be done?

**Comment by antran89** [comment]
Thanks @Aravind-Suresh for your interests. I also do not know where to start. I feel it is better to have a systematical way to help users avoid trivial bugs like this.

**Comment by sturkmen72** [comment]
i think you want to improve [this part of the documentation](http://docs.opencv.org/3.1.0/d1/d10/classcv_1_1MatExpr.html#details). if so you can edit [this part of the source](https://github.com/Itseez/opencv/blob/master/modules/core/include/opencv2/core/mat.hpp#L3248-L3290)

**Comment by antran89** [comment]
Hi @Aravind-Suresh @sturkmen72 
I know it is not so clear for newbie, but I also do not know how to improve it. Thank for your participation.

**Comment by krishraghuram** [comment]
As sturkmen72 said, 
The interface of MatExpr is defined in "/modules/core/include/opencv2/core/mat.hpp"
The implementation of MatExpr is in "/modules/core/src/matop.cpp"
The docs are in the folder "/modules/core/doc"

However, this seems to be limited to C++. Im not able to reproduce the bug in python.

(https://cloud.githubusercontent.com/assets/10024339/26205579/ca110776-3bff-11e7-8fef-a2c655d39b02.png)

But I'm able to reproduce the bug in c++(in the same build of opencv). This bothers me, as this seems to be more than a documentation issue.

**Comment by alalek** [comment]
#7144

There are no implicit type conversions in OpenCV.

**Comment by antran89** [comment]
@alalek I agree that we should have a request for comments from the core developers. 
If we do not support implicit conversions in OpenCV, than we should make the docs clearer for the users for the ease of use.
Thank @alalek for the references.
